mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageResultIterator.java

```diff
@@ -59,36 +59,26 @@
     private MailboxException exception;
     private final Mailbox mailbox;
     private final FetchGroup group;
-    private final MessageUid from;
-    private MessageUid cursor;
-    private final MessageUid to;
+    private final MessageRange range;
     private final BatchSizes batchSizes;
-    private final Type type;
     private final MessageMapper mapper;
     private final FetchType ftype;
 
     public StoreMessageResultIterator(MessageMapper mapper, Mailbox mailbox, MessageRange range, BatchSizes batchSizes, FetchGroup group) {
         this.mailbox = mailbox;
         this.group = group;
         this.mapper = mapper;
-        this.from = range.getUidFrom();
-        this.cursor = this.from;
-        this.to = range.getUidTo();
+        this.range = range;
         this.batchSizes = batchSizes;
-        this.type = range.getType();
         this.ftype = getFetchType(group);
         LOGGER.debug("batchSizes used: {}", batchSizes);
     }
 
     @Override
     public boolean hasNext() {
-        if (cursor.compareTo(to) > 0) {
-            return false;
-        }
-
-        if (next == null || !next.hasNext()) {
+        if (next == null) {
             try {
-                readBatch();
+                next = mapper.findInMailbox(mailbox, range, ftype, batchSizeFromFetchType(ftype));
```

Comment:
As far as I understand, it won't work because if the range is smaller than the batchsize, you need to fetch results several times

Reply:
Good catch